Support for minimum search length configuration parameter.

jimconner opened the pull request:
As mentioned in [issue 2014](https://github.com/mkdocs/mkdocs/issues/2014), searching for any term of less than 3 characters will give the user back a message saying 'No results found' - This is misleading as results may have been present, but due to a hardcoded limit on minimum search length the search was not actually performed.

This PR implements an additional config parameter `min_search_length` for the search plugin.

waylan commented:
I am not comfortable with the message being a "search result". If any such message is being displayed, then it should be in the same format as the "no results" message, which is distinctly different from a search result.

jimconner commented:
I have updated this PR so that it will now accept a min_search_parameter define within the plugin->search block in mkdocs.yml.

I struggled to find a way to get the config values visible within main.js, so I cut 'n' pasted the bits from worker.js that allows it to fetch the config values. Hopefully this is an acceptable approach. I'm more than happy to make tweaks if you can suggest a better approach to take.

jimconner commented:
I see that this is failing to pass the tests in Travis - I'll have a go at fixing that up now.

waylan commented:
> I struggled to find a way to get the config values visible within main.js, so I cut 'n' pasted the bits from worker.js that allows it to fetch the config values. Hopefully this is an acceptable approach.

It is not as it completely negates the benefits of using a worker. Now the json file needs to be fetched twice, once in main and once in the worker. Some users have reported thousands of files which result in huge indexes for their site. Anything except the worker trying to handle that will make their pages unusable.

Also, I find using your product as the example in our documentation to be highly inappropriate.

jimconner commented:
I think there is a misunderstanding here. RabbitMQ is not our product.. just an open-source technology we use that causes our users to search for 'MQ' when they're trying to find out how to use it.

Any pointers on how I can get config values within worker.js would be greatly appreciated. This is my first dabbling with javascript, so I'm mostly trying to guess what the right approach is here. Thanks.

waylan commented:
> Any pointers on how I can get config values within main.js would be greatly appreciated.

You have two options:

1. Include the setting in the json file and have the worker pass it back to the main thread. For example, add a call to `postMessage` right after `init` which passes back config settings:

    https://github.com/mkdocs/mkdocs/blob/2b92209d12bfed050e35eeee2e5073074939727c/mkdocs/contrib/search/templates/search/worker.js#L118-L127

    This would be helpful for other possible future features as well. We haven't done it yet because we haven't yet added a feature which required it.

    As a reminder, we also need to provide a fallback for browsers which do not support WebWorkers here:

    https://github.com/mkdocs/mkdocs/blob/2b92209d12bfed050e35eeee2e5073074939727c/mkdocs/contrib/search/templates/search/main.js#L79-L90

2. Set a variable in the main thread via the template like we do here:

    https://github.com/mkdocs/mkdocs/blob/2b92209d12bfed050e35eeee2e5073074939727c/mkdocs/themes/mkdocs/base.html#L184-L187

    The trick here is that we only want to do that when the search plugin is enabled. And of course, that would add additional requirements for third party themes to support. Therefore, I would prefer the first option to this.

jimconner commented:
Thank you for the pointers Waylan - I'll see if I can get PR attempt 3 arranged tomorrow.

jimconner commented:
Getting close on this now I think. I've got the postMessage mechanism working for passing the config back from worker to main, fixed the failing test, added new one for min_search_length and rebased it to tidy up the commit history. Currently trying to find a way of disabling web workers in either chrome or safari so that I can work on the fallback mechanism. (My google skills are failing me with finding ways to disable WebWorkers)

jimconner commented:
Just about at the end of my working day, so I'll finish off with a question. As I can't work out how to disable WebWorkers in Chrome, I tried forcing the code in main.js to follow the other path...
```
//if (!window.Worker) {
if (true) {
  console.log('Web Worker API not supported');
  // load index in main thread
```

This gives me errors because within the doSearch() function within main.js it attempts `searchWorker.postMessage({query: query});` which cannot succeed because searchWorker is only instantiated in the event that we're using WebWorkers. 

I'm very probably doing something wrong (since this is my first dabbling with javascript), but I thought I should check with you as I don't understand how this could ever work. Is there something else that would be instantiating a searchWorker when WebWorker API is not available ?

waylan commented:
> I tried forcing the code in main.js to follow the other path

My recollection is that I did the same thing when developing the initial implementation. However, there have been some changes by others who are more experienced with JS than me since then, so it's possible that won't work anymore. I'm not sure.

waylan commented:
I just took another look. You need to modify both lines 47 and 81 of `main.js`. Both lines test for `if (!window.Worker)`. Temporarily changing both to `if (true)` should allow you to test the fallback.

jimconner commented:
Thank you for steering me in the right direction once again Waylan - I'll spend some time on this tomorrow to see if I can get it across the line.

jimconner commented:
I've given the alternate code path a test this morning (having changed both to `if (true)` and everything appears to work correctly without requiring any further changes, so I believe that this PR is properly ready for review now. Please let me know of any required changes and I'll (try to) take care of them. Thanks.

waylan commented:
Looks good to me. This can be merged after a note is added to the release notes. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

jimconner commented:
Is adding the note to the release notes something that I need to do? I took a quick look just now but it wasn't obvious where to put in items for upcoming releases.

waylan commented:
Yes, you can add a one-liner here which pints to the original issue.:

https://github.com/mkdocs/mkdocs/blob/653ba22230bc318f9e240f9cf5248c2f41fff410/docs/about/release-notes.md#L24-L27

Note that that was added in a few commits after you started working on this. You may need to rebase with upstream to see those lines in the file.

jimconner commented:
Rebased against upstream/master and added comment into release notes as requested.

waylan commented:
The note in the release notes needs to point to the original issue (#2014). Like this:

    Add support for `min_search_length` parameter for search plugin (#2014).

I would just made the change myself, but it appears you have disabled the ability for us to make changes to your PR and I'm not particularly interesting in forking it and creating another PR.

As an aside, I realize the other two items listed in the current release don't point to issues, but no issues had been opened in each case. This however, does have an issue, so we need to point to it. You can find plenty of examples in previous releases.

jimconner commented:
No idea why the permissions to edit PRs are disabled - Someone else must have set that on our gituib org. I've just force pushed  the fixed release notes with the change that was required.

Thank you so much for all your help and patience getting me across the line with this. Cheers!!
